lib/ipni-client.js

```diff
@@ -45,22 +47,33 @@ export async function queryTheIndex(cid, providerId) {
     const address = p.Provider.Addrs[0]
     if (!address) continue
 
-    switch (protocol) {
-      case 'http':
-        return {
-          indexerResult: 'OK',
-          provider: { address, protocol },
-        }
+    const provider = {
+      address: formatProviderAddress(p.Provider.ID, address, protocol),
+      contextId: p.ContextID,
+      protocol,
+    }
+
+    if (p.Provider.ID === providerId) {
+      switch (protocol) {
+        case 'http':
+          return {
+            indexerResult: 'OK',
+            provider,
+          }
 
-      case 'graphsync':
-        if (!graphsyncProvider) {
-          graphsyncProvider = {
-            address: `${address}/p2p/${p.Provider.ID}`,
-            protocol,
+        case 'graphsync':
+          if (!graphsyncProvider) {
+            graphsyncProvider = provider
           }
-        }
+      }
```

Comment:
It's confusing to me that in case of HTTP we return immediately, but in case of graph sync we don't. I don't see a reason for this behavior. Could we also directly return in the case of graphsync?

Reply:
This was the way we use to do it even before introduction of "alternative providers" to the response. As I understood it: we're iterating trough list of providers and in case we find record serving content via graphsync first we'll return it only if there's no HTTP alternative. In case there's a record serving content via HTTP we return it immediately.